ext/json/json_encoder.c

```diff
@@ -483,6 +483,14 @@ static int php_json_encode_serializable_object(smart_str *buf, zval *val, int op
 	}
 
 	if ((Z_TYPE(retval) == IS_OBJECT) &&
+		(Z_OBJ(retval) != Z_OBJ_P(val)) &&
+		ce->name == Z_OBJCE(retval)->name) {
```

Comment:
Are you trying to compare strings here? `ce->name` is a `zend_string`, so if you want to compare the char values, `==` will only compare the addresses.

Reply:
Nice catch! So if I wanted to compare `zend_string`'s, would I change it to something like this: `ZSTR_VAL(ce->name) == ZSTR_VAL(Z_OBJCE(retval)->name)`?

Reply:
You could use [`zend_string_equals`](https://github.com/php/php-src/blob/master/Zend/zend_string.h#L297), which takes 2 `zend_string *` arguments.  So I believe you could do: `zend_string_equals(ce->name, Z_OBJCE(retval)->name)`

Reply:
Oh awesome - thanks so much! :)

Reply:
@rmccullagh it's a `zend_string *`, `Z_OBJCE` returns a `_zend_class_entry`, and you can find the definition of the structure [here](https://github.com/php/php-src/blob/master/Zend/zend.h#L115). Anyway, I recommend `zend_string_equals` as well. If there is a wrapper method, then use it.

Reply:
Already merged, but I just watched your video so here's some thoughts:

1. If you're just trying to match the class, you can compare the ce only.  This will be more correct than comparing names and (nominally) more performant.  `if ((Z_TYPE_P(retval) == IS_OBJECT) && (Z_OBJCE_P(retval) == ce)) { ... }`

2. The segfault is still easily triggerable via this script. As you mention in your video, there's a much bigger task involved in fixing this properly.

```php
class A implements \JsonSerializable {
  public function jsonSerialize() {
    return new B;
  }
}

class B implements \JsonSerializable {
  public function jsonSerialize() {
    return new A;
  }
}
```

Reply:
@sgolemon this pr has been closed without merge. and the example is exactly what I gave [here](https://github.com/php/php-src/pull/2763#issuecomment-331129755). BTW, what do you mean by video? I don't see any video mentioned here

Reply:
Thanks @sgolemon! I'll update my blog with your comment. :)

@jhdxr I made [a little screencast](https://www.sammyk.me/how-to-find-and-patch-a-bug-in-php-source-php-internals) of how I found & patched this bug. :)

Reply:
@jhdxr Oh, heh. I should really read the conversation first. I was only looking at the code review comments and interpreted the "Closed" as merged because we don't really merge via github. :p